module/VuFind/src/VuFind/Db/Service/ResourceTagsService.php

```diff
@@ -0,0 +1,164 @@
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+
+namespace VuFind\Db\Service;
+
+use Laminas\Paginator\Paginator;
+
+use function count;
+
+/**
+ * Database service for resource_tags.
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+class ResourceTagsService extends AbstractDbService implements
+    ResourceTagsServiceInterface,
+    \VuFind\Db\Table\DbTableAwareInterface
+{
+    use \VuFind\Db\Table\DbTableAwareTrait;
+
+    /**
+     * Get Resource Tags Paginator
+     *
+     * @param ?int    $userId     ID of user (null for any)
+     * @param ?int    $resourceId ID of the resource (null for any)
+     * @param ?int    $tagId      ID of the tag (null for any)
+     * @param ?string $order      The order in which to return the data
+     * @param ?int    $page       The page number to select
+     * @param int     $limit      The number of items to fetch
+     *
+     * @return Paginator
+     */
+    public function getResourceTagsPaginator(
+        ?int $userId = null,
+        ?int $resourceId = null,
+        ?int $tagId = null,
+        ?string $order = null,
+        ?int $page = null,
+        int $limit = 20
+    ): Paginator {
+        return $this->getDbTable('ResourceTags')->getResourceTags($userId, $resourceId, $tagId, $order, $page, $limit);
+    }
+
+    /**
+     * Remove links from the resource_tags table based on an array of IDs.
+     *
+     * @param string[] $ids Identifiers from resource_tags to delete.
+     *
+     * @return int          Count of $ids
+     */
+    public function deleteLinksByResourceTagsIdArray(array $ids): int
+    {
+        return $this->getDbTable('ResourceTags')->deleteByIdArray($ids);
+    }
+
+    /**
+     * Gets unique tagged resources from the database.
+     *
+     * @param ?int $userId     ID of user (null for any)
+     * @param ?int $resourceId ID of the resource (null for any)
+     * @param ?int $tagId      ID of the tag (null for any)
+     *
+     * @return array
```

Comment:
It's not a set of objects, it's just array-formatted data selected from the database. But I guess `array[]` would be more appropriate as a starting point. I haven't gone so far as to document array keys yet; I know there are ways of doing that, but I think they're tool-specific rather than PHP-native, so I have avoided it so far. We may revisit if we start to increase the static analysis requirements in the future.